first_diff.py

```diff
@@ -1,255 +1,55 @@
-    "-c",
-    "--count",
-    type=int,
-    default=5,
-    help="find up to this many instruction difference(s)",
-)
-parser.add_argument(
-    "-d",
-    "--diff",
-    dest="diff_args",
-    nargs="?",
-    action="store",
-    default=False,
-    const="prompt",
-    help="run diff.py on the result with the provided arguments"
-)
-parser.add_argument(
-    "-m", "--make", help="run make before finding difference(s)", action="store_true"
-)
-args = parser.parse_args()
-
-diff_count = args.count
-
-if args.make:
-    check_call(["make", "-j4", "COMPARE=0"])
-
-baseimg = f"baserom.z64"
-basemap = f"expected/build/z64.map"
-
-myimg = f"zelda_ocarina_mq_dbg.z64"
-mymap = f"build/z64.map"
-
-if not os.path.isfile(baseimg):
-    print(f"{baseimg} must exist.")
-    exit(1)
-if not os.path.isfile(myimg) or not os.path.isfile(mymap):
-    print(f"{myimg} and {mymap} must exist.")
+from pathlib import Path
+import rabbitizer
+
+try:
+    import mapfile_parser
+except ImportError:
+    print("Missing dependency mapfile_parser, install it with `python3 -m pip install mapfile-parser`")
     exit(1)
 
-mybin = open(myimg, "rb").read()
-basebin = open(baseimg, "rb").read()
 
-if len(mybin) != len(basebin):
-    print("Modified ROM has different size...")
-    exit(1)
+def decodeInstruction(bytesDiff: bytes, mapFile: mapfile_parser.MapFile) -> str:
+    word = (bytesDiff[0] << 24) | (bytesDiff[1] << 16) | (bytesDiff[2] << 8) | (bytesDiff[3] << 0)
+    instr = rabbitizer.Instruction(word)
+    immOverride = None
+
+    if instr.isJumpWithAddress():
+        # Instruction is a function call (jal)
+
+        # Get the embedded address of the function call
+        symAddress = instr.getInstrIndexAsVram()
+
+        # Search for the address in the mapfile
+        symInfo = mapFile.findSymbolByVramOrVrom(symAddress)
+        if symInfo is not None:
+            # Use the symbol from the mapfile instead of a raw value
+            immOverride = symInfo.symbol.name
+
+    return instr.disassemble(immOverride=immOverride, extraLJust=-20)
+
+def firstDiffMain():
+    parser = argparse.ArgumentParser(description="Find the first difference(s) between the built ROM and the base ROM.")
+
+    parser.add_argument("-c", "--count", type=int, default=5, help="find up to this many instruction difference(s)")
+    parser.add_argument("-v", "--version", help="Which version should be processed", default="mq_dbg")
+    parser.add_argument("-a", "--add-colons", action='store_true', help="Add colon between bytes" )
+
+    args = parser.parse_args()
+
+    buildFolder = Path("build")
+
+    BUILTROM = Path(f"zelda_ocarina_{args.version}.z64")
+    BUILTMAP = buildFolder / f"z64.map"
+
+    EXPECTEDROM = Path("baserom.z64")
+    EXPECTEDMAP = "expected" / BUILTMAP
+
+    exit(mapfile_parser.frontends.first_diff.doFirstDiff(BUILTMAP, EXPECTEDMAP, BUILTROM, EXPECTEDROM, args.count, mismatchSize=True, addColons=args.add_colons, bytesConverterCallback=decodeInstruction))
```

Comment:
Is calling exit() necessary
(I mostly ask because it's not done in sym_info)

Reply:
No, I'll remove it.